### PR TITLE
QCBOREncode_Tell and QCBOREncode_SubString

### PR DESCRIPTION
**Pull request by laurencelundblade**
Address #247 and #248

(initial PR; testing still needed)

### CODE REVIEW

**Review by BrianSipos — COMMENTED** [review]
Even with my comment above, the addition of `QCBOREncode_RetrieveOutputStorage()` and `QCBOREncode_Tell()` fits my needs.

**Comment by BrianSipos** [comment]
This doesn't fully address #248. For the encoder there could be a public API similar to `QCBORDecode_RetrieveInputStorage()` to access the input `UsefulBufC`.

File: inc/qcbor/qcbor_encode.h [hunk 1]
```diff
@@ -2285,6 +2298,68 @@ static QCBORError
...
+ * opened after the call to QCBOREncode_Tell() are closed before this
+ * is called.
+ *
+ * This will return @c NULLUsefulBufC if the encoder is in the error
+ * state or if @c uStart is byeond the end of the thus far encoded
+ * data items.
+ *
+ * See important usage WARNING in QCBOREncode_Tell()
+ */
+UsefulBufC
```
[the remark this answers is not in the document]
Comment (laurencelundblade) [reply inc/qcbor/qcbor_encode.h:2365]: That would be counter to the UsefulBuf convention of UsefulBuf's with data always be const and empty buffers not const. I think that convention is upheld everywhere (including t_cose).

I don't mean to be pedantic about it, but I'd like to keep it as for consistency with the convention.

**Comment by laurencelundblade** [comment]
> This doesn't fully address #248. For the encoder there could be a public API similar to `QCBORDecode_RetrieveInputStorage()` to access the input `UsefulBufC`.

You want the thus-far-encoded CBOR, right? You can get that by passing 0 as the start to SubString(). Finish() can also be called multiple times, but it errors if there are open arrays or maps, where SubString doesn't. (If you look inside the implementation of Finish and SubString are almost the same too)

Thanks for your feedback.

**Comment by laurencelundblade** [comment]
> > This doesn't fully address #248. For the encoder there could be a public API similar to `QCBORDecode_RetrieveInputStorage()` to access the input `UsefulBufC`.
> 
> You want the thus-far-encoded CBOR, right? You can get that by passing 0 as the start to SubString(). Finish() can also be called multiple times, but it errors if there are open arrays or maps, where SubString doesn't. (If you look inside the implementation of Finish and SubString are almost the same too)
> 
> Thanks for your feedback.

Ahh, but you are talking about QCBORDecode() aren't you. That's fine. Got too focused on encode.

**Comment by BrianSipos** [comment]
> Ahh, but you are talking about QCBORDecode() aren't you. That's fine. Got too focused on encode.

Yes it didn't help that I had a typo in my comment. This is about a similar interface for the decoder.

File: inc/qcbor/qcbor_encode.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (BrianSipos) [reply inc/qcbor/qcbor_encode.h:2365]: That's fair and my need can use overall Tell offsets where modifying the encoded form is needed for efficiency.

**Pull request merged**